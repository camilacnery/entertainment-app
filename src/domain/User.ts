export interface IUserProfile {
  imageUrl: string;
  fullName: string;
  firstName: string;
  email: string;
}
