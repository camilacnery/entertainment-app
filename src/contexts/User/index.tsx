import { ReactNode, FC, createContext, useState, useContext, useEffect } from "react";
import { IUserProfile } from "@/domain/User";
import { decodeJwt } from "@/utils/jwt";

interface IDecodedUser {
  aud: string;
  azp: string;
  email: string;
  email_verified: boolean;
  exp: number;
  family_name: string;
  given_name: string;
  iat: number;
  iss: string;
  jti: string;
  name: string;
  nbf: number;
  picture: string;
  sub: string;
}

export type TUserContext = {
  profile?: IUserProfile;
  onSuccessLogin: (userJwt: string) => void;
  onLogout: () => void;
};

const UserContext = createContext<TUserContext>({} as TUserContext);

type Props = {
  children: ReactNode;
};

const STORAGE_ID = "app/user-token";

const UserContextProvider: FC<Props> = ({ children }) => {
  const [profile, setProfile] = useState<IUserProfile>();

  const setProfileFromToken = (userJwt: string) => {
    const profileData = decodeJwt<IDecodedUser>(userJwt);

    setProfile({
      fullName: profileData.name,
      firstName: profileData.given_name,
      email: profileData.email,
      imageUrl: profileData.picture,
    });
  };

  const onSuccessLogin = (userJwt: string) => {
    localStorage.setItem(STORAGE_ID, userJwt);
    setProfileFromToken(userJwt);
  };

  const onLogout = () => {
    localStorage.removeItem(STORAGE_ID);
    setProfile(undefined);
  };

  useEffect(() => {
    const userJwt = localStorage.getItem(STORAGE_ID);
    if (!userJwt) return;

    setProfileFromToken(userJwt);
  }, []);

  return (
    <UserContext.Provider
      value={{
        profile,
        onSuccessLogin,
        onLogout,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

const useUserContext = () => useContext(UserContext);

export { useUserContext, UserContext, UserContextProvider };
