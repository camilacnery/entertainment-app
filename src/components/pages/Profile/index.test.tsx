import { UserContextProvider } from "@/contexts/User";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ProfilePage from ".";

const mockedRouterPush = jest.fn();
jest.mock("next/router", () => ({
  useRouter: () => ({
    push: mockedRouterPush,
  }),
}));

const pageElements = {
  loggedOutError: () =>
    screen.getByRole("heading", {
      name: /you need to be logged in to view this page, redirecting\.\.\./i,
    }),
  logoutButton: () => screen.getByRole("button", { name: /logout/i }),
  profile: {
    name: (name: string) => screen.getByRole("heading", { name }),
    email: (email: string) => screen.getByText(email),
  },
};

const userJwt =
  "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJFbnRlcnRhaW5tZW50IEFwcCIsImlhdCI6MTY2NDc0MzE3MCwiZXhwIjoxNjk2Mjc5MTcwLCJhdWQiOiJ3d3cuZXhhbXBsZS5jb20iLCJzdWIiOiJ0ZXN0QGV4YW1wbGUuY29tIiwiZW1haWwiOiJ0ZXN0QGV4YW1wbGUuY29tIiwiYXpwIjoiaGphc29pa2pscWpubGsiLCJlbWFpbF92ZXJpZmllZCI6InRydWUiLCJmYW1pbHlfbmFtZSI6IlRlc3QiLCJnaXZlbl9uYW1lIjoiVXNlciIsImp0aSI6Impsa2phbHNqZGxhamRsYWoiLCJuYW1lIjoiVGVzdCBVc2VyIiwibmJmIjoiamxja3NhamxkamFzbGQiLCJwaWN0dXJlIjoiaHR0cHM6Ly90ZXN0LnBpY3R1cmUifQ.DHRCtB-XkEcNaULDVjLq_ZHV28ocAYVtHmQLKYBAhmI";

const renderProfile = () => {
  return render(
    <UserContextProvider>
      <ProfilePage />
    </UserContextProvider>
  );
};

describe("Profile Page", () => {
  beforeEach(() => {
    window.localStorage.clear();
  });

  it("renders profile details", () => {
    window.localStorage.setItem("app/user-token", userJwt);
    const { baseElement } = renderProfile();

    expect(baseElement).toMatchSnapshot();
    expect(pageElements.profile.name("Test User")).toBeInTheDocument();
    expect(pageElements.profile.email("test@example.com")).toBeInTheDocument();
  });

  it("clears user data and logout when logout button is clicked", async () => {
    window.localStorage.setItem("app/user-token", userJwt);
    renderProfile();

    await userEvent.click(pageElements.logoutButton());

    expect(window.localStorage.getItem("app/user-token")).toBeNull();
    expect(pageElements.loggedOutError()).toBeInTheDocument();
    expect(mockedRouterPush).toHaveBeenCalledWith("/");
  });

  it("redirects to home when user is not logged in", () => {
    renderProfile();

    expect(pageElements.loggedOutError()).toBeInTheDocument();
    expect(mockedRouterPush).toHaveBeenCalledWith("/");
  });
});
