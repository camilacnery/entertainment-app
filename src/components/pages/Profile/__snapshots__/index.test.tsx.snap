// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Page renders profile details 1`] = `
<body>
  <div>
    <div
      class="banner"
    />
    <main
      class="main"
    >
      <div
        class="content"
      >
        <section
          class="content"
        >
          <img
            alt=""
            class="avatar"
            data-nimg="future"
            decoding="async"
            height="100"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Ftest.picture&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Ftest.picture&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ftest.picture&w=256&q=75 2x"
            style="color: transparent;"
            width="100"
          />
          <h1>
            Test User
          </h1>
          <p>
            test@example.com
          </p>
          <button
            class="button"
          >
            Logout
          </button>
        </section>
      </div>
    </main>
  </div>
</body>
`;
